src/mesh/mesh_dual_weighted_residual_error.cpp

```diff
@@ -0,0 +1,388 @@
+    mpi_communicator(MPI_COMM_WORLD),
+    pcout(std::cout, dealii::Utilities::MPI::this_mpi_process(mpi_communicator)==0)
+{
+    // storing the original FE degree distribution
+    coarse_fe_index.reinit(dg->triangulation->n_active_cells());
+
+    // Create functional
+    functional = FunctionalFactory<dim,nstate,real,MeshType>::create_Functional(dg->all_parameters->functional_param, dg);
+
+    // looping over the cells
+    for(auto cell = dg->dof_handler.begin_active(); cell != dg->dof_handler.end(); ++cell)
+        if(cell->is_locally_owned())
+            coarse_fe_index[cell->active_cell_index()] = cell->active_fe_index();
+}
+
+template <int dim, int nstate, typename real, typename MeshType>
+real DualWeightedResidualError<dim, nstate, real, MeshType>::total_dual_weighted_residual_error(std::shared_ptr< DGBase<dim, real, MeshType> > dg)
+{
+    dealii::Vector<real> cellwise_errors = compute_cellwise_errors(dg);
+    real error_sum = cellwise_errors.l1_norm();
+    return dealii::Utilities::MPI::sum(error_sum, mpi_communicator);
+}
+
+template <int dim, int nstate, typename real, typename MeshType>
+dealii::Vector<real> DualWeightedResidualError<dim, nstate, real, MeshType>::compute_cellwise_errors(std::shared_ptr< DGBase<dim, real, MeshType> > dg)
+{
+    dealii::Vector<real> cellwise_errors(dg->triangulation->n_active_cells());
+    reinit(dg);
+    convert_to_state(AdjointStateEnum::fine, dg);
+    fine_grid_adjoint(dg);
+    cellwise_errors = dual_weighted_residual(dg);
+    convert_to_state(AdjointStateEnum::coarse, dg);
+
+    return cellwise_errors;
+}
+
+
+template <int dim, int nstate, typename real, typename MeshType>
+void DualWeightedResidualError<dim, nstate, real, MeshType>::reinit(std::shared_ptr< DGBase<dim, real, MeshType> > dg)
+{
+    // assuming that all pointers are still valid
+    // reinitilizing all variables after triangulation in the constructor
+    solution_coarse = dg->solution;
+    adjoint_state = AdjointStateEnum::coarse;
+
+    // storing the original FE degree distribution
+    coarse_fe_index.reinit(dg->triangulation->n_active_cells());
+    
+    // looping over the cells
+    for(auto cell = dg->dof_handler.begin_active(); cell != dg->dof_handler.end(); ++cell)
+        if(cell->is_locally_owned())
+            coarse_fe_index[cell->active_cell_index()] = cell->active_fe_index();
+
+    // for remaining, clearing the values
+    dIdw_fine      = dealii::LinearAlgebra::distributed::Vector<real>();
+    dIdw_coarse    = dealii::LinearAlgebra::distributed::Vector<real>();
+    adjoint_fine   = dealii::LinearAlgebra::distributed::Vector<real>();
+    adjoint_coarse = dealii::LinearAlgebra::distributed::Vector<real>();
+
+    dual_weighted_residual_fine = dealii::Vector<real>();
+}
+
+template <int dim, int nstate, typename real, typename MeshType>
+void DualWeightedResidualError<dim, nstate, real, MeshType>::convert_to_state(AdjointStateEnum state, std::shared_ptr< DGBase<dim, real, MeshType> > dg)
```

Comment:
The overall name use of state is confusing a bit since the rest of the code uses state variables, nstate, etc to denote flow/equation/physics variable. Furthermore, you're not really converting the adjoint itself into a fine/coarse adjoints, you're refining the DG solution.

Can you come up with some better function names?

Reply:
It's renamed to indicate that the solution is being refined/coarsened.